src/Processors/Formats/Impl/PrometheusTextOutputFormat.h

```diff
@@ -0,0 +1,51 @@
+#pragma once
+
+#include <Core/Block.h>
+#include <Formats/FormatSettings.h>
+#include <Processors/Formats/IRowOutputFormat.h>
+
+
+namespace DB
+{
+
+class WriteBuffer;
+
+/** A stream for outputting data in tsv format.
+  */
+class PrometheusTextOutputFormat : public IRowOutputFormat
+{
+public:
+    /** with_names - output in the first line a header with column names
+      * with_types - output the next line header with the names of the types
+      */
+    PrometheusTextOutputFormat(
+        WriteBuffer & out_,
+        const Block & header_,
+        const RowOutputFormatParams & params_,
+        const FormatSettings & format_settings_);
+
+    String getName() const override { return "PrometheusTextOutputFormat"; }
+
+    /// https://github.com/prometheus/docs/blob/86386ed25bc8a5309492483ec7d18d0914043162/content/docs/instrumenting/exposition_formats.md
+    String getContentType() const override { return "text/plain; version=0.0.4; charset=UTF-8"; }
+
+protected:
+    void write(const Columns & columns, size_t row_num) override;
+    void writeField(const IColumn &, const ISerialization &, size_t) override {}
+
+
+    struct ColumnPositions
+    {
+        size_t name;
+        size_t value;
+        std::optional<size_t> help;
+        std::optional<size_t> type;
```

Comment:
Maybe also add label support?

I.e.
```
uptime{instance="foo"} 1
uptime{instance="foo"} 2
```

But note, that `type` and `help` can be omitted for metrics that differs only in label, i.e.:

```
# HELP node_softnet_dropped_total Number of dropped packets
# TYPE node_softnet_dropped_total counter
node_softnet_dropped_total{cpu="0"} 0
node_softnet_dropped_total{cpu="1"} 0
```

Reply:
I'm considering options for how to represent labels in the table. We need labels to output `histograms` and `summaries` as well. One option is to store all values for all labels in one row using an array and tuples or maps, like 

```
┌─name──────────────────┬─type──────┬─buckets──────────────────────────────────────────────────────────────────┬──value─┐
│ http_request_duration │ histogram │ [('0.05',24054),('0.1',33444),('0.2',100392),('1',133988),('sum',53423)] │ 144320 │
│ requests              │ counter   │ [('instance="foo"',1),('instance="bar"',2)]                              │   ᴺᵁᴸᴸ │
│ rpc_duration          │ summary   │ [('0.1',3102),('0.5',3272),('0.9',9001),('0.99',14290)]                  │  30292 │
└───────────────────────┴───────────┴──────────────────────────────────────────────────────────────────────────┴────────┘
```
It will be formatted as 
```
# TYPE http_request_duration histogram
http_request_duration_bucket{le="0.05"} 24054
http_request_duration_bucket{le="0.1"} 33444
http_request_duration_bucket{le="0.2"} 100392
http_request_duration_bucket{le="1"} 133988
http_request_duration_bucket{le="+Inf"} 144320
http_request_duration_sum 53423
http_request_duration_count 144320

# TYPE rpc_duration summary
rpc_duration{quantile="0.1"} 3102
rpc_duration{quantile="0.5"} 3272
rpc_duration{quantile="0.9"} 9001
rpc_duration{quantile="0.99"} 14290
rpc_duration_count 30292

# TYPE requests counter
requests{instance="foo"} 1
requests{instance="bar"} 2
```

But it may be difficult to build such arrays for the user using SQL and also timestamp support can be required in addition to value, so it'll be even more complex.

Another option is to use one row for one metric value. In this case, metrics should be grouped. Maybe it's ok to keep requirements about the order of rows and have some state in formatter that would store info about the current group with the same label and format it properly.

Reply:
>But it may be difficult to build such arrays for the user using SQL and also timestamp support can be required in addition to value, so it'll be even more complex.

Indeed, looks complex, but, it will support multiple labels w/o specifying which columns are labels, so maybe it is good enough.

Reply:
@azat
Decided to try to use separate row for each label. Input data should be sorted by metric name. Formatter buffers info about current metric, see test. What do you think about it?